src/clr-angular/data/datagrid/providers/state.provider.ts

```diff
@@ -24,105 +24,94 @@ import { SerializableFilter } from '../interfaces/serializable.filter.interface'
-        if (filter.hasOwnProperty('property') && filter.hasOwnProperty('value')) {
-          const defaultFilterRepresentation = filter as { property: string; value: string };
-          const propertyStringFilter = new DatagridPropertyStringFilter(defaultFilterRepresentation.property);
-          const stringFilter = new DatagridStringFilterImpl(propertyStringFilter);
-          stringFilter.value = defaultFilterRepresentation.value;
-          filterObject = stringFilter;
-        } else {
-          filterObject = filter as SerializableFilter<T>;
+
+    set state(state: ClrDatagridStateInterface<T>) {
+        this.debouncer.changeStart();
+        if (state.page) {
+            this.page.size = state.page.size;
+            this.page.current = Math.ceil(state.page.from / state.page.size) + 1;
         }
-        const existing = activeFilters.findIndex(value => value.compatibleToState(filterObject.filterState));
-        if (existing !== -1) {
-          activeFilters.splice(existing, 1);
-        } else {
-          this.filters.add(filterObject);
+        if (state.sort) {
+            if (typeof state.sort.by === 'string') {
+                if (!(this.sort.comparator instanceof DatagridPropertyComparator)) {
+                    this.sort.comparator = new DatagridPropertyComparator(state.sort.by);
+                }
+                if (this.sort.reverse !== state.sort.reverse) {
+                    this.sort.toggle(this.sort.comparator);
+                }
+            } else {
+                this.sort.comparator = state.sort.by;
+                this.sort.reverse = state.sort.reverse;
+            }
         }
-      }
-      activeFilters.forEach(filter => this.filters.remove(filter));
+        if (state.filters) {
+            const activeFilters = this.filters.getActiveFilters();
+            for (const filter of state.filters) {
+                let filterObject: SerializableFilter<T>;
+                if (filter.hasOwnProperty('property') && filter.hasOwnProperty('value')) {
+                    const defaultFilterRepresentation = filter as { property: string; value: string };
+                    const propertyStringFilter = new DatagridPropertyStringFilter(defaultFilterRepresentation.property);
+                    const stringFilter = new DatagridStringFilterImpl(propertyStringFilter);
+                    stringFilter.value = defaultFilterRepresentation.value;
+                    filterObject = stringFilter;
+                } else {
+                    filterObject = filter as SerializableFilter<T>;
+                }
+                const existing = activeFilters.findIndex(value => value.compatibleToState(filterObject.filterState));
+                if (existing !== -1) {
+                    activeFilters.splice(existing, 1);
+                } else {
+                    this.filters.add(filterObject);
+                }
+            }
+            activeFilters.forEach(filter => this.filters.remove(filter));
```

Comment:
Maybe serializable filters should have an option to just deactivate them instead of removing them. Removing filters also clears all bindings to UI elements, so passing a state without an active string filter and then passing a state with an active string filter probably breaks the whole thing. Did you test this somewhere?

Reply:
it's tested on the UI, it works, but still leaks the memory somewhere.